substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -94,17 +94,246 @@ private PosixJavaNetSubstitutions() {
+            //            return;
+            return;
+        }
+        //        }
+        //        setDefaultScopeID(env, (struct sockaddr *)&him);
+        JavaNetNetUtilMD.setDefaultScopeID(him);
+        //
+        //        if (NET_Bind(fd, (struct sockaddr *)&him, len) < 0)  {
+        //            if (errno == EADDRINUSE || errno == EADDRNOTAVAIL ||
+        //                    errno == EPERM || errno == EACCES) {
+        //                NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "BindException",
+        //                                             "Bind failed");
+        //            } else {
+        //                NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+        //                                             "Bind failed");
+        //            }
+        //            return;
+        //        }
+        if (JavaNetNetUtilMD.NET_Bind(fd, him, len_Pointer.read()) < 0) {
+            if (Errno.errno() == Errno.EADDRINUSE() || Errno.errno() == Errno.EADDRNOTAVAIL() || //
+                    Errno.errno() == Errno.EPERM() || Errno.errno() == Errno.EACCES()) {
+                throw new BindException(PosixUtils.lastErrorString("Bind failed"));
+            } else {
+                throw new SocketException(PosixUtils.lastErrorString("Bind failed"));
+            }
+        }
+        //
+        //    /* initialize the local port */
+        //        if (localport == 0) {
+        //        /* Now that we're a connected socket, let's extract the port number
+        //         * that the system chose for us and store it in the Socket object.
+        //         */
+        //            if (JVM_GetSockName(fd, (struct sockaddr *)&him, &len) == -1) {
+        //                NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+        //                                             "Error getting socket name");
+        //                return;
+        //            }
+        //
+        //            localport = NET_GetPortFromSockaddr((struct sockaddr *)&him);
+        //
+        //            (*env)->SetIntField(env, this, pdsi_localPortID, localport);
+        //        } else {
+        //            (*env)->SetIntField(env, this, pdsi_localPortID, localport);
+        //        }
+        if (lport == 0) {
+            if (VmPrimsJVM.JVM_GetSockName(fd, him, len_Pointer) == -1) {
+                throw new SocketException(PosixUtils.lastErrorString("Error getting socket name"));
+            }
+            // 602         localport = NET_GetPortFromSockaddr((struct sockaddr *)&him);
+            lport = JavaNetNetUtilMD.NET_GetPortFromSockaddr(him);
+            // 603         (*env)->SetIntField(env, this, psi_localportID, localport);
+            Util_java_net_PlainDatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).localPort = lport;
```

Comment:
Would it be worth having a `Util_java_net_PlainDatagramSocketImpl.setLocalPort(...., int)` to mimic this `SetIntField`, including mimicking the Java class hierarchy where it is missing from the `Target` class hierarchy?  I almost suggested up where you get the `fd` field.  It has to be a static method taking `this` as a parameter, but that just makes the declaration ugly, not so much the use.

(Wow, I hope the C and Java compilers do "common tail" optimizations to pull that assignment out of both the `then` and `else` branches.  Not my problem, this week. :-)